scripts/schemas/board-schema.yml

```diff
@@ -78,3 +78,56 @@ mapping:
     type: seq
     sequence:
       - include: board-schema
+  run_once:
+    type: seq
+    desc: |
+          Allows for restricting west flash commands when using sysbuild to run once per given
+          grouping of boards. This is to allow for future image program cycles to not erase the
+          flash of a device which has just been programmed by another image.
+    sequence:
+      - type: map
+        mapping:
+          regex;(.*):
+            required: true
+            type: seq
+            desc: |
+                  A list of commands which should be limited to running once per invocation of
+                  west flash for a set of boards
+            sequence:
+              - type: map
+                mapping:
+                  boards:
+                    required: true
+                    type: seq
+                    desc: |
+                          A grouping of boards which the command should apply to. Can be used
+                          multiple times to have multiple groups.
+                    sequence:
+                      - type: str
```

Comment:
This seems to allow users to refer to any board, regardless of this board being described in this yaml or not.

For example:
```
board:
  name: plank
  socs:
  - name: soc1
run_once:
  - '--a-command':
     - boards:
        - 'foo'
```

It should not be possible to reference boards beside those defined by this board.yml.

Reply:
We have a huge problem with `west boards` being incredibly slow because of schema validation, and to add extra overhead on top of this of then doing regex matching every time it is invoked does not really make sense. Verification as part of compliance in CI makes sense but whenever someone runs `west flash` no

Reply:
>  and to add extra overhead on top of this of then doing regex

are your really arguing that a regex like this is faster ?
```
        mapping:
          regex;(.*):
```

then I would like to see some performance numbers.

Reply:
I still lack some performance numbers.

Also, this principle has high risk of wrongly indenting yaml code, for example:
```
    - '--recover':
      run: first
      runners:
      - nrfjprog
      - jlink
```
is wrong, but this works:
```
    - '--recover':
            run: first
            runners:
            - nrfjprog
            - jlink
```
This minor detail of remembering the extra indentation is extremely important but may not be obvious to everyone, and the schema validation failure is not helpful at all.

Thus, I still argue that a normal key-value pair is much better, in this case like this:
```
    - command: '--recover':
      run: first
      runners:
      - nrfjprog
      - jlink
```

No surprises to users on how to do indentation.

Reply:
as for performance, the first reason for `west boards` is the split in almost 500 boards.yml files, which all need to be validated, not whether you try to sub-optimize the key.

In fact, the overhead of `west` itself is quite significant.
Running `west boards --board foo` gives: (`--board foo` is to avoid time overhead of printing)
```
$ time west boards --board foo
_______________________________________________________
Executed in  715.04 millis    fish           external
   usr time  698.32 millis    0.00 micros  698.32 millis
   sys time   16.36 millis  304.00 micros   16.05 millis
```

but removing `west` and use `list_boards.py` directly, give you:
```
time ./scripts/list_boards.py --board-root=(pwd) --arch-root=(pwd) --soc-root=(pwd) --board=foo
________________________________________________________
Executed in  543.97 millis    fish           external
   usr time  514.78 millis    0.00 micros  514.78 millis
   sys time   28.28 millis  350.00 micros   27.93 millis
```

Reply:
> Also, this principle of using values for keys through the regex has high risk of users wrongly indenting yaml code, for example:
> This minor detail of remembering the extra indentation is extremely important but may not be obvious to everyone, and the schema validation failure is not helpful at all.

The user won't get into that situation because the schema would fail

Reply:
> The user won't get into that situation because the schema would fail

any user creating a oot board has the risk of falling into this.

And any Zephyr contributor can waste a lot of time on this.

Reply:
Then I suggest you improve the error reporting by submitting a PR to https://pypi.org/project/pykwalify/